modules/handlers/command_handlers.py

```diff
@@ -290,3 +291,140 @@ def forwarded_post_msg(update: Update, context: CallbackContext):
+
+    abusive_message_id = MemeData.get_last_post_report(user_id=info['sender_id'], channel_id=channel_id)
+
+    info['bot'].forward_message(chat_id=chat_id,
+                                from_chat_id=channel_id,
+                                message_id=abusive_message_id)
+
+    info['bot'].sendMessage(chat_id=chat_id, text="🚨🚨 SEGNALAZIONE 🚨🚨\n\n" + info['text'])
+    info['bot'].send_message(chat_id=info['chat_id'],
+                            text="Gli admins verificheranno quanto accaduto. Grazie per la collaborazione!")
+
+    return STATE['end']
+
+# endregion
+
+# region report user
+
+def report_cmd(update: Update, context: CallbackContext) -> int:
+    """Handles the reply to the key "Report".
+    Checks the message the user wants to report
+
+    Args:
+        update (Update): update event
+        context (CallbackContext): context passed by the handler
+
+    Returns:
+        int: next state of the conversation
+    """
+    info = get_message_info(update, context)
+    if update.message.chat.type != "private":  # you can only post with a private message
+        info['bot'].send_message(
+            chat_id=info['chat_id'],
+            text="Non puoi usare quest comando ora\nChatta con @Spotted_DMI_bot in privato")
+        return STATE['end']
+
+    user_report = MemeData.get_user_report(user_id=info['sender_id'])
+
+    if user_report:
+        delta_time = datetime.now() - datetime.strptime(user_report['message_date'], "%Y-%m-%d %H:%M:%S.%f")
+        delta_minutes = delta_time.total_seconds()/60
+        remain_minutes = int(config_map['meme']['report_wait_mins'] - delta_minutes)
+
+        if context.user_data['report_sent'] and remain_minutes > 0:
+            info['bot'].send_message(chat_id=info['chat_id'], text=f"Aspetta {remain_minutes} minuti.")
+            return STATE['end']
+
+    info['bot'].send_message(chat_id=info['chat_id'], text="Invia l'username di chi vuoi segnalare. Es. @massimobene")
+
+    return STATE['reporting_user']
+
+def report_user_msg(update: Update, context: CallbackContext) -> int:
+    """Handles the reply to the /report command after sent the @username.
+    Checks the the user wants to report, and goes to ask the reason
+
+    Args:
+        update (Update): update event
+        context (CallbackContext): context passed by the handler
+
+    Returns:
+        int: next state of the conversation
+    """
+    info = get_message_info(update, context)
+    if not check_message_type(update.message) or not info['text'].startswith('@') or info['text'].find(' ') != -1:  # the type is NOT supported
+        info['bot'].send_message(
+            chat_id=info['chat_id'],
+            text="Questo tipo di messaggio non è supportato\nÈ consentito solo username telegram. Puoi annullare il processo con /cancel")
+        return STATE['reporting_user']
+
+    MemeData.set_user_report(user_id=info['sender_id'], evil_username=info['text'])
+    context.user_data['report_sent'] = False
```

Comment:
I think the line
```python
MemeData.set_user_report(user_id=info['sender_id'], evil_username=info['text'])
 ```
should be removed from here, since it is present in *report_user_sent_msg*, where it belongs, and if left here would insert a record in the database even if the user later uses /cancel.
My suggestion was to use the *context.user_data* this way:
```python
context.user_data['current_report_target'] = info['text']
```
so in *report_user_sent_msg* you could easily retrive the *evil_username* and only then store the finalized report in the database.